File: java/application/src/main/java/dev/ufadevconf/application/PickupPointApplicationService.java
package dev.ufadevconf.application;

import dev.ufadevconf.domain.Parcel;
import dev.ufadevconf.domain.PickupPoint;
import dev.ufadevconf.domain.PickupPointRepository;
import dev.ufadevconf.domain.Weight;
import lombok.RequiredArgsConstructor;

import java.util.Objects;

@RequiredArgsConstructor
public class PickupPointApplicationService {

    private final PickupPointRepository pickupPointRepository;

    public PickupPoint registerPickupPoint(CreatePickupPointCommand command) {
        Objects.requireNonNull(command, "command");
        var pickupPoint = new PickupPoint(
                command.code(),
                command.address(),
                new Weight(command.minWeightGrams()),
                new Weight(command.maxWeightGrams())
        );
        pickupPointRepository.save(pickupPoint);
        return pickupPoint;
    }

    public ParcelFitResult checkParcelFits(CheckParcelFitsQuery query) {
        Objects.requireNonNull(query, "query");
        var pickupPoint = pickupPointRepository.findByCode(query.pickupPointCode())
                .orElseThrow(() -> new PickupPointNotFoundException(query.pickupPointCode()));
        var parcel = new Parcel(
                query.trackingNumber(),
                new Weight(query.parcelWeightGrams()),
                query.contentsDescription()
        );
        return new ParcelFitResult(
                pickupPoint.getCode(),
                parcel.trackingNumber(),
                pickupPoint.canAccept(parcel)
        );
    }
}

File: kotlin/application/src/main/kotlin/dev/ufadevconf/application/PickupPointApplicationService.kt
package dev.ufadevconf.application

import dev.ufadevconf.domain.Parcel
import dev.ufadevconf.domain.PickupPoint
import dev.ufadevconf.domain.PickupPointRepository
import dev.ufadevconf.domain.Weight

class PickupPointApplicationService(
    private val pickupPointRepository: PickupPointRepository
) {
    fun registerPickupPoint(command: CreatePickupPointCommand): PickupPoint {
        val pickupPoint = PickupPoint(
            code = command.code,
            address = command.address,
            minWeight = Weight(command.minWeightGrams),
            maxWeight = Weight(command.maxWeightGrams)
        )
        pickupPointRepository.save(pickupPoint)
        return pickupPoint
    }

    fun checkParcelFits(query: CheckParcelFitsQuery): ParcelFitResult {
        val pickupPoint = pickupPointRepository.findByCode(query.pickupPointCode)
            ?: throw PickupPointNotFoundException(query.pickupPointCode)
        val parcel = Parcel(
            trackingNumber = query.trackingNumber,
            weight = Weight(query.parcelWeightGrams),
            contentsDescription = query.contentsDescription
        )
        return ParcelFitResult(
            pickupPointCode = pickupPoint.code,
            trackingNumber = parcel.trackingNumber,
            fits = pickupPoint.canAccept(parcel)
        )
    }
}
